docs/DesignDocs/AttestationQPLEnforcementSupport.md

```diff
@@ -0,0 +1,95 @@
+Attestation SGX QPL Enforcement Support
+====
+
+This proposal is to make a change to `oe_verify_evidence`, so that it can support policy-based SGX QPL (Quote Provider Library) specific enhancement, such as allowing the API user to pass data to the SGX QPL for more advanced quote verification.
+
+Motivation
+----
+
+Intel SGX rolls out PCK certificate and collateral information about every six months.
+There is a scenario where a quote signed by the last version of PCK certificate cannot be verified with the latest version of collateral information.
+This causes unnecessary attestation failures when new versions of PCK certificate and collateral information are rolled out.
+To deal with this issue, the SGX QPL need to return any version of collateral information based on additional query parameters (referred to as "baselines").
+With that, a new API called `sgx_ql_get_quote_verification_collateral_with_params` was introduced in the SGX QPL.
+Changes are also needed in the OE SDK to call this new API.
+
+Goals:
+ - Verifier can call `oe_verify_evidence` and pass SGX QPL-specific parameters to do more advanced quote verification.
+
+User Experience
+----
+
+A new policy type `OE_POLICY_SGX_COLLATERAL_BASELINE` is added to current policy types.
+A quote verifier application can pass multiple policies with this type, while OE SDK would pass all the policies down to the SGX QPL API without interpreting the data content.
+
+Specification
+----
+
+### New policy type & its data structure
+
+- `OE_POLICY_SGX_COLLATERAL_BASELINE` in `oe_policy_type_t`
+- `oe_query_param`, which is defined as below,
+```C
+// defined by SGX QPL
+#define MAX_PARAM_STRING_SIZE (255)
+typedef struct _oe_query_param
+{
+uint8_t key[MAX_PARAM_STRING_SIZE+1];
+uint8_t value[MAX_PARAM_STRING_SIZE+1];
+} oe_query_param_t;
+```
+
+### New APIs / OCalls
+
+Though there is no API signature changes for `oe_verify_evidence`, some internal APIs that potentially get used by tools need to be changed.
+For backward compatibility, the following new APIs will be added to ensure all existing code can still work,
+- `oe_get_sgx_endorsements_with_policies`, after change, this API will be called by existing `oe_get_sgx_endorsements` without any policy.
+This new API will not be exposed to OE SDK users.
+- `oe_get_quote_verification_collateral_with_params_ocall`, the current API is `oe_get_quote_verification_collateral_ocall` is an API defined in sgx/attestation.edl, which could have been used by user enclaves, so need to bring this new API to work with existing API.
+
+### Code example
+
+With the new policy, the OE SDK client will be able to pass in multiple collateral baseline parameters as below,
```

Comment:
Where should the OE SDK client get the list of legal key names/values from?  I.e., what list is "region" and "tcbevaluationdatanumber" on?   Can we link to a reference here?

Reply:
Good question, however, the answer is we don't have a list. It's specifically defined by SGX, and as of now, region, tcbevaluationdatanumber are the keys that I got from them. I would prefer the list to be maintained somewhere in SGX official document site, however, I can update this doc with the link when it's ready, but as of now, the official list is not yet ready.

Reply:
Ok, I just think the API is not really usable without a referenceable list.

Reply:
that's true, let me put a placeholder here for future update